### PR TITLE
Add outlier plot

### LINKED ISSUE #348: Add plot for identifying outliers
The current numeric profiles don't show an accurate representation of the data density. A rug plot or a kernel density plot can be added for identifying outliers more easily.

Part of #342

### PR DESCRIPTION
**Pull request by djbarnwal**
Added a plot in numeric profiles to highlight outliers.
Current approach is based on using a higher resolution histogram (500 buckets being used) for creating a rug plot.
Closes #348 

Preview:
image

### CODE REVIEW

File: src/lib/components/viz/histogram/OutlierHistogram.svelte [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+      <feMerge>
+        <feMergeNode in="OUTLINE" />
+        <feMergeNode in="SourceGraphic" />
+      </feMerge>
+    </filter>
+    <g class="textElements">
+        {#each intervals as [label, value] (label)}
+            <line
+                x1={x(value)}
```
Comment (bcolloran) [inline src/lib/components/viz/histogram/OutlierHistogram.svelte:73]: i'm seeing type errors for this use of value, can you update that?-- 
image

**Review by bcolloran — CHANGES_REQUESTED** [review]
on thing i've noticed is that in the case of data with very dramatic outliers, the  std dev labels end up overlapping:
image

I think it's fine if you want to merge now and think about that more later, or you can look at it more now if you'd like.

I did notice one small typing issue, once that is fixed you can merge as far as I'm concerned. Nice work Dhiraj!

File: src/lib/components/viz/histogram/OutlierHistogram.svelte (hunk 1 above)
Comment (djbarnwal) [reply src/lib/components/viz/histogram/OutlierHistogram.svelte:73]: Interesting, VSCode is not throwing any error for me. 
`x1` has type `(property) svelte.JSX.SVGAttributes<SVGLineElement>.x1?: string | number` when I inspect it. The `OutlierHistogram` component is derived from the `NumericHistogram` component. Do you see a similar type error there as well?

**Comment by djbarnwal** [comment]
> on thing i've noticed is that in the case of data with very dramatic outliers, the std dev labels end up overlapping: 

Yes, missed this case. I think overlapping labels are a big issue. Any thoughts on how should we handle this design wise? @magorlick @bcolloran 

Would it be better if I just hide the deviation labels in those cases or alternatively just display the 3sigma label?

**Comment by bcolloran** [comment]
> Would it be better if I just hide the deviation labels in those cases or alternatively just display the 3sigma label?

I'm not really sure. I was also conflicted about whether it would be helpful to to show the sd of the most extreme outlier, but maybe it doesn't matter if its like 243sigma (this is the problem with outliers...)

I wonder if perhaps we should step back and really consider what the objective of this plot is? First, let me say that your implementation is great @djbarnwal-- but as I think more about this and look at the plot in the context of the product, I'm not sure it's actually adding much. In the case of data without big outliers (like the screenshots of winddir and humidity at the top) IMO it doesn't add much above and beyond the histograms we already have for showing the distribution -- it's like a noisier, less clear histogram. And in the case of data with severe outliers (e.g. the screenshots I added)... it also doesn't add much.

@magorlick what's your take? What user story are we trying to solve for? Do we want to make the distribution clear, or highlight outliers? (Sounds like the same thing, but not always the case.) If we want to clarify the distribution, we might do something like only showing the central 99.9% (or whatever), and throwing away outliers; or perhaps add a "show/hide outliers" toggle. Or, to clarify the prevalence of outliers, something like leaderboard might actually be better -- just enumerate the outliers.

Sorry to go meta on this one, just not sure there's a great solution at the level of SD labels (which, also as I reflect more on it, convey similar info the min/max/IQR labels the histogram already has...). No problem if we want to continue on this path and just tun up the labels, I just don't have any good ideas on that path at this point.

**Comment by magorlick** [comment]
The point of the rug plot is to better highlight outliers and extremes that melt into the unseen edges of the histograms and other weird bin artifacts.  

I really liked the SD labels because they give you a better sense of exactly how extreme things are, but it does look a little strange when you have a super extreme outlier and 3SD is stacked on top of the mean.  My preference is to keep these labels and show the SD of the most extreme outlier and the mean is instead of 1,2,3 SD when there and points more extreme than 3 SD away.

In the future, it might be nice to add some transparency to the rug plot so we can get a better sense of the density of points when lots of points overlap as seen in cases like this.

**Comment by bcolloran** [comment]
I guess I want to continue a friendly interrogation of some of these ideas to make sure we're using pixels wisely and really getting at the underlying need for our users

To this point:
> the rug plot is to better highlight outliers and extremes that melt into the unseen edges of the histograms and other weird bin artifacts

Looking at the those screen shots above, I'm not convinced the rugplot is really doing this better than the existing min/max labels. In the case of distributions without outliers, the bins do seem to adequately capture the extremes, but in the case of the distributions that do have outliers, we end up with a single lonely tick mark for the outlier (whose presence is already explicitly called out by the max/min label), and an indistinct clump of ticks melding into a single solid rectangular block of pixels for the main body of the distribution.

Because of this mooshing into an indistinct clump, it feels like perhaps the transparency may be really needed for the rug plot to convey added info.

For now: I think I'd choose to pass on this feature (which, again, is not due to execution, which i very good, just to the reality of seeing it) because to me it spends down our pixel budget without recouping much new value, but I'm really fine to be outvoted if other folks are finding value in the approach.

(All of this questioning is lightly held, and, again: I'm happy to have it be out of scope for now and go ahead to merge this. But for context: I built a tool literally called the "Distribution Viewer" at Mozilla, and I spent a lot of time thinking about how to depict 1d distributions and handle outliers, and I'm recalling some of that now. When focused on the task of helping to the user to understand the shape of actual data distribution, not displaying outliers was the best default. When focused on the task of helping people to identify outliers, visual display (on linear x-axis -- log x-axis can help, but brings its own challenges) may not be right because it can actually end up hiding the shape of the data when there are extreme values (as in the screenshot). Maybe more to think about for a future rev.)

File: src/lib/components/viz/histogram/OutlierHistogram.svelte (hunk 1 above)
Comment (bcolloran) [reply src/lib/components/viz/histogram/OutlierHistogram.svelte:73]: For me the missing type was on `value`, which could be of type `number | string`. Looking farther up the file, I think `intervals` need a tuple type. This resolved it for me:

`const intervals:[string, number][] = [["µ", mean]];` (line 32)

Adding that clears the missing types on all those instanes of `value`.

If your VScode is not highlighting those, ping me on slack or something -- I can share my VSCode config to make sure we're seeng the same stuff ! :-)

**Comment by djbarnwal** [comment]
Just FYI, this is how it would look like when handling extreme outliers

image

Also to the point of adding transparency based on count/frequency, wouldn't it be just another histogram where the scale is now opacity instead of height of the bar? 

Also to Brendan's point, I do think we can create another model where we can have a toggle to hide outliers. Like for the above example <1% of data points are deviating the shape of the histogram. We can use deviations or some statistical methods to remove those outliers for the profile. Again, I am not sure how the tradeoffs work here and what direction would be best to move forward with.

**Comment by magorlick** [comment]
I like the extreme outlier labels with the mean in your last mock.  Could you do a time-boxed effort to play with adding 'transparency' to the rug plot.  Instead of actually adding transparency, you might try a range of colors from pink to red if this is a quick win.  Identifying the rarer outliers and clusters are a big part of the value add for the rug plot so we don't to make it so close to the background color that you can't see small outliers. One note is that if we go with pink to red instead of a transparency we will need to revisit our approach if we create a dark-mode of RD in the future.  Not a big concern right now.

**Comment by djbarnwal** [comment]
Adding transparency is proving to be a bit harder than I thought. We have made [this](https://github.com/rilldata/rill-developer/pull/357) change in order to reduce the number of elements in the DOM. So essentially now the Histogram is just 1 element with the shape sketched out as we can play with the coordinates of the plotted line.

In the old implementation we were using individual elements so we could freely control their transparency. The new implementation is faster but getting transparency to work would be tough on that one big element (just like tooltips or interactions present in timeseries profiles are harder to replicate in numeric profiles). There could be some solution using `clip` but I am not aware of it's mechanics or if it works for our case.

**Comment by magorlick** [comment]
@djbarnwal Thank you for looking into it, let's pass on the transparency for now and wrap up this work.

**Review by bcolloran — APPROVED** [review]

**Pull request merged**